Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift

```diff
@@ -727,9 +731,521 @@ internal final class _CalendarGregorian: _CalendarProtocol, @unchecked Sendable
+            // this month is the cutover month
+            return (attemptDayOfMonth, monthStartDate, monthLen - kCutoverMonthRollbackDays, true)
+        }
+
+        return (day, monthStartDate, monthLen, false)
+    }
+
+    // Limits the day of month to the number of days in the given month
+    func capDay(in dateComponent: inout DateComponents) {
+        guard let day = dateComponent.day, let month = dateComponent.month, let year = dateComponent.year else {
+            return
+        }
+
+        let daysInMonth = numberOfDaysInMonth(month, year: year)
+        if day > daysInMonth {
+            dateComponent.day = daysInMonth
+        } else if day < 1 {
+            dateComponent.day = 1
+        }
+    }
+
+    func add(_ field: Calendar.Component, to date: Date, amount: Int, inTimeZone timeZone: TimeZone) -> Date {
+
+        let amountInSeconds: Int
+        var nanoseconds: Double = 0
+
+        let allComponents: Calendar.ComponentSet = [.era, .year, .month, .day, .hour, .minute, .second, .nanosecond, .weekday, .weekdayOrdinal, .quarter, .weekOfMonth, .weekOfYear, .yearForWeekOfYear]
+        switch field {
+        case .era:
+            // We've been ignorning era historically. Do the same for compatibility reason.
+            return date
+
+        case .yearForWeekOfYear:
+            var dc = dateComponents(allComponents, from: date, in: timeZone)
+            var amount = amount
+            if let era = dc.era, era == 0 {
+                amount = -amount
+            }
+            dc.yearForWeekOfYear = (dc.yearForWeekOfYear ?? 0) + amount
+            capDay(in: &dc)
+            let resolvedComponents = ResolvedDateComponents(preferComponent: .yearForWeekOfYear, dateComponents: dc)
+            let old = self.date(from: dc, inTimeZone: timeZone, resolvedComponents: resolvedComponents)!
+            return old
+        case .year:
+            var dc = dateComponents(allComponents, from: date, in: timeZone)
+            var amount = amount
+            if let era = dc.era, era == 0 {
+                amount = -amount
+            }
+            dc.year = (dc.year ?? 0) + amount
+            capDay(in: &dc)
+            let resolvedComponents = ResolvedDateComponents(preferComponent: .year, dateComponents: dc)
+            return self.date(from: dc, inTimeZone: timeZone, resolvedComponents: resolvedComponents)!
+
+        case .month:
+            var dc = dateComponents(allComponents, from: date, in: timeZone)
+            dc.month = (dc.month ?? 0) + amount
+            capDay(in: &dc) // adding 1 month to Jan 31 should return Feb 29, not Feb 31
+            let resolvedComponents = ResolvedDateComponents(preferComponent: .month, dateComponents: dc)
+            return self.date(from: dc, inTimeZone: timeZone, resolvedComponents: resolvedComponents)!
+
+        case .quarter:
+            // TODO: This isn't supported in Calendar_ICU either. We should do it here though.
```

Comment:
I wonder if part of the reason this is unsupported (for adding, at least) is that it may be unclear what adding "1 quarter" to a date might mean. Is it the same as 3 months? I think it's far more likely to use in matching where you want to know when the start of some specific quarter is, for example.

Reply:
I don't think it's an unreasonable API though. It's as useful as figuring out the date when someone says, "Let's revisit the plan **next quarter**". I think it's reasonable to treat adding one quarter as adding three months, but whether we want to preserve the day of the month or the weekday is open to discussion.